File: src/infinicore/ops/blas_dot/blas_dot.cc
#include "infinicore/ops/blas_dot.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

common::OpDispatcher<BlasDot::schema> &BlasDot::dispatcher() {
    static common::OpDispatcher<BlasDot::schema> dispatcher_;
    return dispatcher_;
};

void BlasDot::execute(Tensor result, Tensor x, Tensor y) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(result, x, y);
    infinicore::context::setDevice(result->device());
    dispatcher().lookup(result->device().getType())(result, x, y);
}

Tensor blas_dot(Tensor x, Tensor y) {
    auto result = Tensor::empty({}, x->dtype(), x->device());
    blas_dot_(result, x, y);
    return result;
}

void blas_dot_(Tensor result, Tensor x, Tensor y) {
    BlasDot::execute(result, x, y);
}

} // namespace infinicore::op

File: python/infinicore/ops/blas_dot.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def blas_dot(x: Tensor, y: Tensor, *, out=None):
    if out is None:
        return Tensor(_infinicore.blas_dot(x._underlying, y._underlying))

    _infinicore.blas_dot_(x._underlying, y._underlying, out._underlying)

    return out
